sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -327,121 +329,311 @@ bool WinHttpTransport::VerifyCertificatesInChain(
   }
-  catch (std::exception& ex)
+  void WinHttpAction::CompleteActionWithError(DWORD_PTR stowedErrorInformation, DWORD stowedError)
   {
-    // If an exception is thrown in the handler, log the error and terminate the connection.
-    Log::Write(Logger::Level::Error, "Exception Thrown: " + std::string(ex.what()));
+    std::unique_lock<std::mutex> lock(m_actionCompleteMutex);
+    m_stowedErrorInformation = stowedErrorInformation;
+    m_stowedError = stowedError;
+    SetEvent(m_actionCompleteEvent.get());
   }
-}
 
-/**
- * @brief HTTP Callback to enable private certificate checks.
- *
- * This method is called by WinHTTP when a certificate is received. This method is called multiple
- * times based on the state of the TLS connection. We are only interested in
- * WINHTTP_CALLBACK_STATUS_SENDING_REQUEST, which is called during the TLS handshake.
- *
- * When called, we verify that the certificate chain sent from the server contains the certificate
- * the HTTP client was configured with. If it is, we accept the connection, if it is not,
- * we abort the connection, closing the incoming request handle.
- */
-void WinHttpTransport::OnHttpStatusOperation(HINTERNET hInternet, DWORD dwInternetStatus)
-{
-  if (dwInternetStatus != WINHTTP_CALLBACK_STATUS_SENDING_REQUEST)
+  DWORD WinHttpAction::GetStowedError()
+  {
+    std::unique_lock<std::mutex> lock(m_actionCompleteMutex);
+    return m_stowedError;
+  }
+  DWORD_PTR WinHttpAction::GetStowedErrorInformation()
   {
-    if (dwInternetStatus == WINHTTP_CALLBACK_STATUS_SECURE_FAILURE)
+    std::unique_lock<std::mutex> lock(m_actionCompleteMutex);
+    return m_stowedErrorInformation;
+  }
+  DWORD WinHttpAction::GetBytesAvailable()
+  {
+    std::unique_lock<std::mutex> lock(m_actionCompleteMutex);
+    return m_bytesAvailable;
+  }
+
+  /**
+   * Called by WinHTTP when sending a request to the server. This callback allows us to inspect
+   * the TLS certificate before sending it to the server.
+   */
+  void WinHttpAction::StatusCallback(
+      HINTERNET hInternet,
+      DWORD_PTR dwContext,
+      DWORD internetStatus,
+      LPVOID statusInformation,
+      DWORD statusInformationLength) noexcept
+  {
+    // If we're called before our context has been set (on Open and Close callbacks), ignore the
+    // status callback.
+    if (dwContext == 0)
     {
-      Log::Write(Logger::Level::Error, "Security failure. :(");
+      return;
     }
-    // Silently ignore if there's any statuses we get that we can't handle
-    return;
-  }
 
-  // We will only set the Status callback if a root certificate has been set.
-  AZURE_ASSERT(!m_options.ExpectedTlsRootCertificates.empty());
+    try
+    {
+      WinHttpAction* httpAction = reinterpret_cast<WinHttpAction*>(dwContext);
+      httpAction->OnHttpStatusOperation(
+          hInternet, internetStatus, statusInformation, statusInformationLength);
+    }
+    catch (Azure::Core::RequestFailedException& rfe)
```

Comment:
```suggestion
    catch (Azure::Core::RequestFailedException const& rfe)
```

Reply:
Is there any benefit to making this a const reference? Does it change the generated code in any way?

I'm not a huge fan of adding const to variable declarations unless there are specific semantic benefits (for instance, declaring a parameter as const means that the caller knows that the underlying function won't modify the value of the parameter so it can cache values across the method call). I'm not aware of a similar benefit for local variables.

Reply:
Deleted previous comment after digging in a bit.